resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -100,11 +97,56 @@ protected DnsNameResolverContext(DnsNameResolver parent,
         allowedQueries = maxAllowedQueries;
     }
 
-    protected Promise<T> promise() {
-        return promise;
+    void resolve(Promise<T> promise) {
+        boolean directSearch = parent.searchDomains().length == 0 || StringUtil.endsWith(hostname, '.');
+        if (directSearch) {
+            internalResolve(promise);
+        } else {
+            final Promise<T> original = promise;
+            promise = parent.executor().newPromise();
+            promise.addListener(new FutureListener<T>() {
+                @Override
+                public void operationComplete(Future<T> future) throws Exception {
+                    if (future.isSuccess()) {
+                        original.setSuccess(future.getNow());
+                    } else {
+                        future.addListener(new FutureListener<T>() {
+                            int count;
+                            @Override
+                            public void operationComplete(Future<T> future) throws Exception {
+                                if (future.isSuccess()) {
+                                    original.trySuccess(future.getNow());
+                                } else {
+                                    if (count < parent.searchDomains().length) {
+                                        String searchDomain = parent.searchDomains()[count++];
+                                        Promise<T> nextPromise = parent.executor().newPromise();
+                                        String nextHostname = DnsNameResolverContext.this.hostname + "." + searchDomain;
+                                        DnsNameResolverContext<T> nextContext = newResolverContext(parent,
+                                            nextHostname, resolveCache);
+                                        nextContext.internalResolve(nextPromise);
+                                        nextPromise.addListener(this);
+                                    } else {
+                                        original.tryFailure(future.cause());
+                                    }
+                                }
+                            }
+                        });
+                    }
+                }
+            });
+            int idx = hostname.length();
+            int dots = 0;
+            while (idx-- > 0) {
+                if (hostname.charAt(idx) == '.' && ++dots >= parent.ndots()) {
+                    internalResolve(promise);
```

Comment:
sure. I'll just provide the code bcz it gets the point across :) We don't need to create a new promise, add a listener, etc... if we never find the number of expected dots.

```java
    void resolve(Promise<T> promise) {
        boolean directSearch = parent.searchDomains().length == 0 ||
            (hostname.length() > 0 && hostname.charAt(hostname.length() - 1) == '.');
        if (directSearch) {
            internalResolve(promise);
        } else {
            int dots = 0;
            for (int i = hostname.length() - 1; i >= 0; --i) {
                if (hostname.charAt(i) == '.' && ++dots >= parent.ndots()) {
                    internalResolve(promise);
                    final Promise<T> original = promise;
                    promise = parent.executor().newPromise();
                    promise.addListener(new FutureListener<T>() {
                        @Override
                        public void operationComplete(Future<T> future) throws Exception {
                            if (future.isSuccess()) {
                                original.setSuccess(future.getNow());
                            } else {
                                future.addListener(new FutureListener<T>() {
                                    private int count;
                                    @Override
                                    public void operationComplete(Future<T> future) throws Exception {
                                        if (future.isSuccess()) {
                                            original.trySuccess(future.getNow());
                                        } else {
                                            if (count < parent.searchDomains().length) {
                                                String searchDomain = parent.searchDomains()[count++];
                                                Promise<T> nextPromise = parent.executor().newPromise();
                                                String nextHostname = DnsNameResolverContext.this.hostname + "." + searchDomain;
                                                DnsNameResolverContext<T> nextContext = newResolverContext(parent,
                                                        nextHostname, resolveCache);
                                                nextContext.internalResolve(nextPromise);
                                                nextPromise.addListener(this);
                                            } else {
                                                original.tryFailure(future.cause());
                                            }
                                        }
                                    }
                                });
                            }
                        }
                    });
                    return;
                }
            }

            promise.tryFailure(new UnknownHostException(hostname));
        }
    }
```